python/taulabs/telemetry.py

```diff
@@ -212,36 +268,139 @@ def __select(self, finishTime):
+
+            if written > 0:
+                self.send_buf = self.send_buf[written:]
 
             didStuff = True
 
         return didStuff
 
-    def __receive(self, finishTime):
-        """ Fetch available data from TCP socket """
+    def _send(self, msg):
+        """ Send a string out the TCP socket """
 
-        # Always do some minimal IO if possible
-        self.__select(0)
+        self.send_buf += msg
 
-        while (len(self.recv_buf) < 1) and self.__select(finishTime):
-            pass
+        self._do_io(0)
 
-        if len(self.recv_buf) < 1:
-            return None
+    def _done(self):
+        return self.fd is None
 
-        ret=self.recv_buf
-        self.recv_buf=''
+class NetworkTelemetry(FDTelemetry):
+    def __init__(self, host="127.0.0.1", port=9000, *args, **kwargs):
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.connect((host, port))
 
-        return ret
+        s.setblocking(0)
+
+        self.sock = s
+
+        FDTelemetry.__init__(self, fd=s.fileno(), *args, **kwargs)
+
+class FileTelemetry(TelemetryBase):
+    # TODO accept file object
+    def __init__(self, filename='sim_log.tll', parseHeader=False,
+             *args, **kwargs):
+        self.filename = filename
+        self.f = file(filename, 'r')
+
+        if parseHeader:
+            # Check the header signature
+            #    First line is "Tau Labs git hash:"
+            #    Second line is the actual git hash
+            #    Third line is the UAVO hash
+            #    Fourth line is "##"
+            sig = self.f.readline()
+            if sig != 'Tau Labs git hash:\n':
+                print "Source file does not have a recognized header signature"
+                print '|' + sig + '|'
+                raise IOError("no header signature")
+            # Determine the git hash that this log file is based on
+            githash = self.f.readline()[:-1]
+            if githash.find(':') != -1:
+                import re
+                githash = re.search(':(\w*)\W', githash).group(1)
+
+            print "Log file is based on git hash: %s" % githash
+
+            uavohash = self.f.readline()
+            divider = self.f.readline()
+
+            TelemetryBase.__init__(self, serviceInIter=False, iterBlocks=True,
+                doHandshaking=False, githash=githash, *args, **kwargs)
+        else:
+            TelemetryBase.__init__(self, serviceInIter=False, iterBlocks=True,
+                doHandshaking=False, *args, **kwargs)
+
+        self.done=False
+        self.start_thread()
+
+    def _receive(self, finishTime):
+        """ Fetch available data from file """
+
+        buf = self.f.read(128)
+
+        if buf == '':
+            self.done=True
+
+        return buf
+
+    def _done(self):
+        return self.done
+
+def _normalize_path(path):
+    import os
+    return os.path.normpath(os.path.join(os.getcwd(), path))
+
+def GetUavoBasedOnArgs(desc=None):
```

Comment:
I thought about this.  argparse kinda assumes it's doing it all or doing none-- there's not an easy way if a parent called it to get the stuff that didn't match args (unless I missed it).   If they really care, they can instantiate FileTelemetry directly, I thought.